plugins/primus_lisp/semantics/bits.lisp

```diff
@@ -19,13 +19,87 @@
          (logand (msb rm) (lnot (msb rd)))
          (logand (msb rn) (lnot (msb rd)))))
 
-
 (defun overflow (rd rn rm)
   "(overflow RD RN RM) is true if the sum RD = RN + RM results
    in two's complement overflow."
   (logor (logand (msb rn) (msb rm) (lnot (msb rd)))
          (logand (lnot (msb rn)) (lnot (msb rm)) (msb rd))))
 
+(defun highest-set-bit (bitv)
+  "(highest-set-bit bitv) returns the greatest index whose bit is set in bitv.
+   It requires bitv to be non-zero.
+   Translated from ARMv8 ISA pseudocode."
+  (assert-msg (not (is-zero bitv)) "highest-set-bit bitv is zero")  ; at least 1 bit must be set
+  (let ((i (- (word-width bitv) 1)))
+    (while (and (> i 0) (= (select i bitv) 0))
+      (decr i))
+    i))
+
+(defun replicate (bitv n)
+  "(replicate bitv n) returns a bitvector with bitv repeated n times.
+   Translated from ARMv8 ISA pseudocode."
+  (let ((output 0:0))
```

Comment:
while this works it is on the edge of Primus Lisp semantics. Strictly speaking, `0:0` is an ill-formed bitvector and it is ill-typed (the set of zero-width bitvectors is uninhabited). Eventually, this will  be fixed in the implementation and this code will stop working. So my suggestion is to rewrite it so that the initial `output` is equal to `bitv` with the updated loop condition `(> n 1)`

Reply:
done 👍